### PR TITLE
Migrate `run-make/issue-46239` to `rmake`

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

### CODE REVIEW

File: tests/run-make/no-mark-noalias/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// In the rustc_trans crate, references returned by a Frozen pointer type
```
Comment (jieyouxu) [inline tests/run-make/no-mark-noalias/rmake.rs:1]: This comment is outdated, we don't have a `rustc_trans` crate anymore do we?

File: tests/run-make/no-mark-noalias/rmake.rs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+// could mark them as "noalias", which caused miscompilation errors.
```
Comment (jieyouxu) [inline tests/run-make/no-mark-noalias/rmake.rs:2]: I feel this should be a codegen test if we're trying to check for `noalias` metadata?

**Comment by jieyouxu** [comment]
I'll revisit this, this seems like it should be a codegen test

**Comment by jieyouxu** [comment]
See [codegen tests](https://rustc-dev-guide.rust-lang.org/tests/compiletest.html#codegen-tests) for more references, please do ask me if you are stuck or have any questions.

@rustbot author

**Comment by Oneirical** [comment]
I have attempted the conversion to a codegen test, more so we can discuss it than claim "it's done". Seeing the other tests putting their flags on library functions and not `main()` makes me a little worried (I took inspiration from other noalias-related codegen tests such as `noalias-box.rs`.)

@rustbot review

File: tests/codegen/noalias-freeze.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+fn project<T>(x: &(T,)) -> &T { &x.0 }
+
+fn dummy() {}
+
+// CHECK-NOT: noalias
+fn main() {
+    let f = (dummy as fn(),);
+    (*project(&f))();
+}
```
Comment (jieyouxu) [inline tests/codegen/noalias-freeze.rs:17]: Suggestion: I think this test could be written as

```rs
//@ compile-flags: -Copt-level=1

#![crate_type = "lib"]

fn project<T>(x: &(T,)) -> &T { &x.0 }

fn dummy() {}

// CHECK-LABEL: @foo(
// CHECK-NOT: noalias
#[no_mangle]
pub fn foo() {
    let f = (dummy as fn(),);
    (*project(&f))();
}
```

This way the noalias check must happen after the `foo` label occurs.

**Comment by Oneirical** [comment]
@rustbot review

Thanks for the suggestion! I'm just now learning how the codegen tests work, and I see how you made it more similar to the other codegen tests, and prevented the name `foo` from being altered by compilation with `no-mangle`.

**Comment by jieyouxu** [comment]
Thanks, r=me after squashing commits into one.
@bors delegate+

**Pull request merged**